File: src/cmds/node.ts
import { run } from '../core/runner';
import { stripAnsi, truncate } from '../core/utils';
import { HandlerResult } from './git';

type PM = 'npm' | 'pnpm' | 'yarn';

export function handleNode(pm: PM, args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run(pm, args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  let cmdType = `${pm} ${sub || 'cmd'}`;

  try {
    if (sub === 'install' || sub === 'i' || sub === 'add' || (pm === 'yarn' && sub === '')) {
      filtered = filterInstallOutput(raw, pm, ultra);
    } else if (sub === 'list' || sub === 'ls') {
      filtered = filterList(raw, ultra);
    } else if (sub === 'outdated') {
      filtered = filterOutdated(raw, ultra);
    } else if (sub === 'run' || sub === 'run-script' || (pm === 'yarn' && sub !== 'install')) {
      filtered = filterRun(raw, result.exitCode, ultra);
    } else {
      filtered = truncate(stripAnsi(raw).trim(), 30);
    }
  } catch {
    filtered = raw;
  }

  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : (result.stderr || raw).trim()),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType,
    execMs: result.execMs,
  };
}

function filterInstallOutput(raw: string, pm: PM, ultra = false): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n');

  let total = 0;
  // npm/pnpm patterns
  const m1 = /added\s+(\d+)\s+package/.exec(clean);
  const m2 = /(\d+)\s+packages? installed/.exec(clean);
  const m3 = /Done in [\d.]+s/.exec(clean);
  const m4 = /Progress:.*?([\d]+)\/([\d]+)/.exec(clean);
  const m5 = /\+\s+(\S+)@(\S+)/g;
  if (m1) total = parseInt(m1[1], 10);
  else if (m2) total = parseInt(m2[1], 10);

  const direct: string[] = [];
  let dm: RegExpExecArray | null;
  while ((dm = m5.exec(clean)) !== null) {
    direct.push(`${dm[1]}@${dm[2]}`);
  }
  for (const line of lines) {
    const a = /^added\s+(\S+)@(\S+)/.exec(line.trim());
    if (a) direct.push(`${a[1]}@${a[2]}`);
  }

  const errors: string[] = [];
  const warnings: string[] = [];
  for (const line of lines) {
    if (/^npm ERR!|^pnpm ERR|^error /i.test(line)) errors.push(line.trim());
    if (/^npm warn|^pnpm WARN|^warning /i.test(line)) warnings.push(line.trim());
  }
  const uniqueWarnings = Array.from(new Set(warnings)).slice(0, 5);
  const uniqueDirect = Array.from(new Set(direct)).slice(0, 10);

  if (errors.length > 0) {
    if (ultra) return `✗${errors.length}err`;
    return `✗ Install failed\n${errors.slice(0, 5).join('\n')}`;
  }

  if (ultra) {
    const tag = total > 0 ? `✓${total}` : '✓ok';
    if (uniqueDirect.length > 0) {
      const compact = uniqueDirect
        .map((d) => d.replace(/@(\d+)\.\d+\.\d+/, '@$1'))
        .slice(0, 3)
        .join(' ');
      return `${tag} ${compact}`;
    }
    return tag;
  }

  const out: string[] = [];
  if (total > 0) out.push(`✓ Installed ${total} packages`);
  else if (m3) out.push('✓ Done');
  else out.push('✓ ok');
  if (uniqueDirect.length > 0) {
    out.push('');
    out.push(`New: ${uniqueDirect.join(', ')}`);
  }
  if (uniqueWarnings.length > 0) {
    out.push('');
    out.push(`Warnings (${uniqueWarnings.length}):`);
    for (const w of uniqueWarnings) out.push(`  ${w}`);
  }
  return out.join('\n');
}

function filterList(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => /\S/.test(l));
  const top: string[] = [];
  for (const line of lines) {
    if (/^[├└]/.test(line) || /^\S+@/.test(line)) {
      if (!/[│ ]{2}/.test(line)) top.push(line.trim());
    }
  }
  if (ultra) return `${top.length} pkgs`;
  if (top.length === 0) return truncate(clean, 30);
  return `Top-level dependencies (${top.length}):\n${top.slice(0, 30).join('\n')}`;
}

function filterOutdated(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const lines = clean.split('\n').filter((l) => /\S/.test(l));
  const dataLines = lines.slice(1).filter((l) => !/^Package/i.test(l));
  if (ultra) return `${dataLines.length} outdated`;
  if (dataLines.length === 0) return '✓ All up-to-date';
  return `${dataLines.length} outdated packages\n${dataLines.slice(0, 20).join('\n')}`;
}

function filterRun(raw: string, exitCode: number, ultra: boolean): string {
  const clean = stripAnsi(raw);
  if (exitCode === 0) {
    if (ultra) return '✓';
    const lastLines = clean.split('\n').filter((l) => l.trim()).slice(-5).join('\n');
    return lastLines || '✓ ok';
  }
  if (ultra) return '✗';
  return truncate(clean, 30);
}

File: internal/handlers/node.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

// HandleNode compresses npm / pnpm / yarn output.
func HandleNode(pm string, args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run(pm, args)
	raw := combined(r)
	filtered := ""

	switch {
	case sub == "install" || sub == "i" || sub == "add" || (pm == "yarn" && sub == ""):
		filtered = filterInstall(raw, ultra)
	case sub == "list" || sub == "ls":
		filtered = filterList(raw, ultra)
	case sub == "outdated":
		filtered = filterOutdated(raw, ultra)
	case sub == "run" || sub == "run-script" || (pm == "yarn" && sub != "install"):
		filtered = filterRun(raw, r.ExitCode, ultra)
	default:
		filtered = util.Truncate(strings.TrimSpace(util.StripAnsi(raw)), 30)
	}

	cmdType := pm + " " + sub
	if sub == "" {
		cmdType = pm + " cmd"
	}
	return finalize(filtered, r, raw, cmdType)
}

var (
	addedPkgsRe = regexp.MustCompile(`added\s+(\d+)\s+package`)
	pkgsInstRe  = regexp.MustCompile(`(\d+)\s+packages? installed`)
	doneInRe    = regexp.MustCompile(`Done in [\d.]+s`)
	plusPkgRe   = regexp.MustCompile(`\+\s+(\S+)@(\S+)`)
	addedLineRe = regexp.MustCompile(`^added\s+(\S+)@(\S+)`)
	npmErrRe    = regexp.MustCompile(`(?i)^npm ERR!|^pnpm ERR|^error `)
	npmWarnRe   = regexp.MustCompile(`(?i)^npm warn|^pnpm WARN|^warning `)
	majorVerRe  = regexp.MustCompile(`@(\d+)\.\d+\.\d+`)
	nonEmptyRe  = regexp.MustCompile(`\S`)
	listTopRe   = regexp.MustCompile(`^[├└]|^\S+@`)
	listNestRe  = regexp.MustCompile(`[│ ]{2}`)
	pkgHeaderRe = regexp.MustCompile(`(?i)^Package`)
)

func filterInstall(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	lines := strings.Split(clean, "\n")

	total := 0
	if m := addedPkgsRe.FindStringSubmatch(clean); m != nil {
		total = atoi(m[1])
	} else if m := pkgsInstRe.FindStringSubmatch(clean); m != nil {
		total = atoi(m[1])
	}

	var direct []string
	for _, m := range plusPkgRe.FindAllStringSubmatch(clean, -1) {
		direct = append(direct, m[1]+"@"+m[2])
	}
	for _, line := range lines {
		if a := addedLineRe.FindStringSubmatch(strings.TrimSpace(line)); a != nil {
			direct = append(direct, a[1]+"@"+a[2])
		}
	}

	var errs, warns []string
	for _, line := range lines {
		if npmErrRe.MatchString(line) {
			errs = append(errs, strings.TrimSpace(line))
		}
		if npmWarnRe.MatchString(line) {
			warns = append(warns, strings.TrimSpace(line))
		}
	}
	uniqueWarnings := limit(util.Unique(warns), 5)
	uniqueDirect := limit(util.Unique(direct), 10)

	if len(errs) > 0 {
		if ultra {
			return fmt.Sprintf("✗%derr", len(errs))
		}
		return "✗ Install failed\n" + strings.Join(limit(errs, 5), "\n")
	}

	if ultra {
		tag := "✓ok"
		if total > 0 {
			tag = fmt.Sprintf("✓%d", total)
		}
		if len(uniqueDirect) > 0 {
			var compact []string
			for _, d := range limit(uniqueDirect, 3) {
				compact = append(compact, majorVerRe.ReplaceAllString(d, "@$1"))
			}
			return tag + " " + strings.Join(compact, " ")
		}
		return tag
	}

	var out []string
	switch {
	case total > 0:
		out = append(out, fmt.Sprintf("✓ Installed %d packages", total))
	case doneInRe.MatchString(clean):
		out = append(out, "✓ Done")
	default:
		out = append(out, "✓ ok")
	}
	if len(uniqueDirect) > 0 {
		out = append(out, "", "New: "+strings.Join(uniqueDirect, ", "))
	}
	if len(uniqueWarnings) > 0 {
		out = append(out, "", fmt.Sprintf("Warnings (%d):", len(uniqueWarnings)))
		for _, w := range uniqueWarnings {
			out = append(out, "  "+w)
		}
	}
	return strings.Join(out, "\n")
}

func filterList(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	var top []string
	for _, line := range strings.Split(clean, "\n") {
		if !nonEmptyRe.MatchString(line) {
			continue
		}
		if listTopRe.MatchString(line) && !listNestRe.MatchString(line) {
			top = append(top, strings.TrimSpace(line))
		}
	}
	if ultra {
		return fmt.Sprintf("%d pkgs", len(top))
	}
	if len(top) == 0 {
		return util.Truncate(clean, 30)
	}
	return fmt.Sprintf("Top-level dependencies (%d):\n%s", len(top), strings.Join(limit(top, 30), "\n"))
}

func filterOutdated(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	var lines []string
	for _, l := range strings.Split(clean, "\n") {
		if nonEmptyRe.MatchString(l) {
			lines = append(lines, l)
		}
	}
	var dataLines []string
	for i, l := range lines {
		if i == 0 || pkgHeaderRe.MatchString(l) {
			continue
		}
		dataLines = append(dataLines, l)
	}
	if ultra {
		return fmt.Sprintf("%d outdated", len(dataLines))
	}
	if len(dataLines) == 0 {
		return "✓ All up-to-date"
	}
	return fmt.Sprintf("%d outdated packages\n%s", len(dataLines), strings.Join(limit(dataLines, 20), "\n"))
}

func filterRun(raw string, exitCode int, ultra bool) string {
	clean := util.StripAnsi(raw)
	if exitCode == 0 {
		if ultra {
			return "✓"
		}
		var nonEmpty []string
		for _, l := range strings.Split(clean, "\n") {
			if strings.TrimSpace(l) != "" {
				nonEmpty = append(nonEmpty, l)
			}
		}
		if last := lastN(nonEmpty, 5); len(last) > 0 {
			return strings.Join(last, "\n")
		}
		return "✓ ok"
	}
	if ultra {
		return "✗"
	}
	return util.Truncate(clean, 30)
}
